### PR TITLE
introduce CoreServices to decouple backend from QWidgets

### PR DESCRIPTION
**Pull request by Be-ing**
This is a prerequisite for using a QQuickWindow (or QOpenGLWindow) as Mixxx's GUI.

Inspired by #941.

TODO:
  * [x] cleanup sloppy usage of std::shared_ptr::get

### CODE REVIEW

File: src/coreservices.cpp [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+    // effect backends to refer to controls that are produced by the engine.
+    BuiltInBackend* pBuiltInBackend = new BuiltInBackend(m_pEffectsManager.get());
+    m_pEffectsManager->addEffectsBackend(pBuiltInBackend);
+#ifdef __LILV__
+    m_pLV2Backend = new LV2Backend(m_pEffectsManager.get());
+    // EffectsManager takes ownership
+    m_pEffectsManager->addEffectsBackend(m_pLV2Backend);
+#else
+    m_pLV2Backend = nullptr;
+#endif
```
Comment (Be-ing) [inline src/coreservices.cpp:203]: Note that this mess is cleaned up in #2618

File: src/mixxx.cpp [hunk 2]
```diff
@@ -1489,15 +932,19 @@ void MixxxMainWindow::rebootMixxxView() {
 }
 
 bool MixxxMainWindow::loadConfiguredSkin() {
+    // TODO: use std::shared_ptr throughout skin widgets instead of these hacky get() calls
```
Comment (Be-ing) [inline src/mixxx.cpp:935]: This would require changes in tons of files throughout the GUI code. Let's leave it for another PR.

**Comment by Be-ing** [comment]
Please review and merge quickly. This changes lots of files and will develop merge conflicts quickly.

File: src/coreservices.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    TooltipsPreference m_toolTipsCfg;
+
+    Timer m_runtime_timer;
+    const CmdlineArgs& m_cmdlineArgs;
+
+    ScreenSaverPreference m_inhibitScreensaver;
+
+    QSet<ControlObject*> m_skinCreatedControls;
+
+    static const int kMicrophoneCount;
```
Comment (uklotzde) [inline src/coreservices.h:139]: constepr

**Review by uklotzde — COMMENTED** [review]
How about using typedefs for all thosee pointer types instead of scattering std::shared_ptr everywhere in the code?

**Comment by uklotzde** [comment]
Thank you for taking over! I agree that we should reduce the changes in this initial step to a bar minimum. Just split the classes and hand out managed pointers.

**Comment by Be-ing** [comment]
> How about using typedefs for all thosee pointer types instead of scattering std::shared_ptr everywhere in the code?

I'm not really a fan of this pattern. If the number of characters in the type name are a concern, I'd rather use `using namespace std` than typedef everything.

**Comment by uklotzde** [comment]
> > How about using typedefs for all thosee pointer types instead of scattering std::shared_ptr everywhere in the code?
> 
> I'm not really a fan of this pattern. If the number of characters in the type name are a concern, I'd rather use `using namespace std` than typedef everything.

Ok, I also don't have a strong opinion about this. `std::shared_ptr<ClassName>` is transparent and unambiguous.

**Review by daschuer — CHANGES_REQUESTED** [review]
Some comments.

File: src/preferences/dialog/dlgpreferences.h [hunk 4]
```diff
@@ -4,12 +4,13 @@
...
 #include <QStringList>
+#include <memory>
 
-#include "preferences/dialog/ui_dlgpreferencesdlg.h"
-#include "preferences/usersettings.h"
 #include "control/controlpushbutton.h"
+#include "preferences/dialog/ui_dlgpreferencesdlg.h"
 #include "preferences/dlgpreferencepage.h"
 #include "preferences/settingsmanager.h"
+#include "preferences/usersettings.h"
```
[the remark this answers is not in the document]
Comment (Be-ing) [reply src/preferences/dialog/dlgpreferences.h:13]: Please don't ask me to change code I did not touch. I am keeping this PR as small as possible to get it merged ASAP.

File: src/mixxx.cpp [hunk 5]
```diff
@@ -82,338 +82,48 @@
...
-            &MixxxMainWindow::slotNoVinylControlInputConfigured);
-    PlayerInfo::create();
-
-    for (int i = 0; i < kMicrophoneCount; ++i) {
-        m_pPlayerManager->addMicrophone();
-    }
-
-    for (int i = 0; i < kAuxiliaryCount; ++i) {
-        m_pPlayerManager->addAuxiliary();
+    for (const auto& group : m_pCoreServices->getPlayerManager()->getVisualPlayerGroups()) {
```
[the remark this answers is not in the document]
Comment (Be-ing) [reply src/mixxx.cpp:123]: I don't understand what you mean. This is a QStringList that is being iterated.

[the remark this answers is not in the document]
Comment (Holzhaus) [reply src/mixxx.cpp:123]: Can you replace `auto` with `QString` to make that obvious?

[the remark this answers is not in the document]
Comment (daschuer) [reply src/mixxx.cpp:123]: Good Idea. 

I meant original: 
DEBUG_ASSERT(m_pCoreServices && m_pCoreServices->getPlayerManager())

**Comment by Be-ing** [comment]
I don't know how to deal with this clazy error:
```
Error: /home/runner/work/mixxx/mixxx/src/mixxx.cpp:123:5: error: c++11 range-loop might detach Qt container (QStringList) [-Wclazy-range-loop]
    for (const auto& group : m_pCoreServices->getPlayerManager()->getVisualPlayerGroups()) {
    ^
```
Neither `std::as_const` nor `qAsConst` work with QStringList

File: src/mixxx.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/mixxx.cpp:123]: Please don't suggest to combine multiple asserts with `&&`. Use separate statements instead.

[the remark this answers is not in the document]
Comment (Be-ing) [reply src/mixxx.cpp]: okay 
6fee01f

**Comment by uklotzde** [comment]
> I don't know how to deal with this clazy error:
> 
> ```
> Error: /home/runner/work/mixxx/mixxx/src/mixxx.cpp:123:5: error: c++11 range-loop might detach Qt container (QStringList) [-Wclazy-range-loop]
>     for (const auto& group : m_pCoreServices->getPlayerManager()->getVisualPlayerGroups()) {
>     ^
> ```
> 
> Neither `std::as_const` nor `qAsConst` work with QStringList

I remember that Jan introduced temporary variables for these cases.

File: src/mixxx.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply src/mixxx.cpp]: Asserting that m_pCoreServices isn't null would be pointless. The application would have crashed before that if it was.

File: src/preferences/dialog/dlgpreferences.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (daschuer) [reply src/preferences/dialog/dlgpreferences.h:13]: OK. I don't mind. But it is only a matter of removing a line.

File: src/mixxx.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (daschuer) [reply src/mixxx.cpp]: Yes, so the assert can be placed at the very beginning of the function.

**Review by daschuer — APPROVED** [review]

File: src/mixxx.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply src/mixxx.cpp]: That seems superfluous, but okay, done: 
2b1eb59

**Comment by Be-ing** [comment]
CI passed. Merge?

**Comment by uklotzde** [comment]
LGTM

**Comment by Be-ing** [comment]
Thanks everyone for being quick with this review so it actually got merged without building up tons of conflicts.

File: src/mixxx.cpp [hunk 6]
```diff
@@ -82,338 +82,52 @@
...
-    FontUtils::initializeFonts(resourcePath); // takes a long time
-
-    launchProgress(2);
-
-    // Set the visibility of tooltips, default "1" = ON
-    m_toolTipsCfg = static_cast<mixxx::TooltipsPreference>(
-        pConfig->getValue(ConfigKey("[Controls]", "Tooltips"),
-                static_cast<int>(mixxx::TooltipsPreference::TOOLTIPS_ON)));
-
-    m_pTouchShift = new ControlPushButton(ConfigKey("[Controls]", "touch_shift"));
```
Comment (Holzhaus) [inline src/mixxx.cpp:238]: This PR was merged prematurely. I now get debug assertions on `main`:
```
warning [Main] ControlDoublePrivate::getControl returning NULL for ( "[Controls]" , "touch_shift" )
critical [Main] DEBUG ASSERT: "flags.testFlag(ControlFlag::NoAssertIfMissing)" in function static QSharedPointer<ControlDoublePrivate> ControlDoublePrivate::getControl(const ConfigKey&, ControlFlags, ControlObject*, bool, bool, bool, double) at /home/jan/Projects/mixxx/src/control/control.cpp:151
critical [Main] DEBUG ASSERT: "m_pControl || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /home/jan/Projects/mixxx/src/control/controlproxy.cpp:37
critical [Main] DEBUG ASSERT: "valid() || flags.testFlag(ControlFlag::NoAssertIfMissing)" in function void ControlProxy::initialize(ControlFlags) at /home/jan/Projects/mixxx/src/control/controlproxy.cpp:38
```

Comment (Be-ing) [reply src/mixxx.cpp:238]: fixed in #3453

**Comment by uklotzde** [comment]
Another bug appeared: Tool tips are now enabled on every restart. Anyone else experiences that?

File: src/coreservices.h [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    std::shared_ptr<PlayerManager> getPlayerManager() const {
+        return m_pPlayerManager;
+    }
+
+    std::shared_ptr<RecordingManager> getRecordingManager() const {
+        return m_pRecordingManager;
+    }
+
```
Comment (jospezial) [inline src/coreservices.h:61]: I think it needs #ifdef __BROADCAST__
with -DBROADCAST=off:
In file included from /var/tmp/portage/media-sound/mixxx-9999/work/mixxx-9999/src/mixxx.h:8,
                 from /var/tmp/portage/media-sound/mixxx-9999/work/mixxx-9999/src/mixxx.cpp:1:
/var/tmp/portage/media-sound/mixxx-9999/work/mixxx-9999/src/coreservices.h: In member function ‘std::shared_ptr<BroadcastManager> mixxx::CoreServices::getBroadcastManager() const’:
/var/tmp/portage/media-sound/mixxx-9999/work/mixxx-9999/src/coreservices.h:64:16: error: ‘m_pBroadcastManager’ was not declared in this scope; did you mean ‘getBroadcastManager’?
   64 |         return m_pBroadcastManager;
      |                ^~~~~~~~~~~~~~~~~~~
      |                getBroadcastManager

Comment (daschuer) [reply src/coreservices.h:61]: Yes. Right, I will take care.

Comment (daschuer) [reply src/coreservices.h:61]: https://github.com/mixxxdj/mixxx/pull/3514

**Comment by ywwg** [comment]
It looks like this PR is the source of a bunch of regressions, is there a plan to fix them?

**Comment by Be-ing** [comment]
What are they?

**Comment by ronso0** [comment]
tooltips visibility & vinyl lead-in time not saved, a few skin COs not available for controllers
